onboarding tasks

Gjjjiang opened the pull request:
# Purpose
Explain the purpose of the PR here, including references to any existing Notion tasks.

# New Changes
- Explain new changes

# Testing
- Explain tests that you ran to verify code functionality.

# Outstanding Changes
- If there are non-critical changes (i.e. additional features) that can be made to this feature in the future, indicate them here.

Gjjjiang committed:
onboarding tasks

dgobalak commented on onboarding/source/amb_light_service.c:
```diff
@@ -18,22 +26,91 @@ static void lightServiceTask(void * pvParameters);
+    if (lightServiceQueueHandle == NULL) {
+        lightServiceQueueHandle = xQueueCreate(LIGHT_SERVICE_QUEUE_SIZE,
+                                                LIGHT_SERVICE_ITEM_SIZE);
+    }
```
Need to check if the queue was created successfully.

Gjjjiang replied:
Done

dgobalak commented on onboarding/source/amb_light_service.c:
```diff
@@ -18,22 +26,91 @@ static void lightServiceTask(void * pvParameters);
+            while(!adcIsConversionComplete(adcREG1, adcGROUP1));
+            adcGetSingleData(adcREG1, adcGROUP1, lightDataPtr);
+            char *msg = malloc(20 * sizeof(char));
+            sprintf(msg, "light level: %d", lightData.value);
```
`snprintf` is preferred over `sprintf` since it's more secure. It ensures you don't exceed the buffer size.

Gjjjiang replied:
Done

dgobalak commented on onboarding/source/amb_light_service.c:
```diff
@@ -18,22 +26,91 @@ static void lightServiceTask(void * pvParameters);
+            adcStartConversion_selChn(adcREG1, LIGHT_SENSOR_PIN, 1, adcGROUP1);
+            while(!adcIsConversionComplete(adcREG1, adcGROUP1));
+            adcGetSingleData(adcREG1, adcGROUP1, lightDataPtr);
+            char *msg = malloc(20 * sizeof(char));
```
We tend to avoid dynamically allocated memory wherever possible since we want our system to be deterministic. Since `malloc` can fail if we run out of space, it introduces nondeterministic behaviour.

dgobalak replied:
Change `msg` to a stack-allocated array.

dgobalak replied:
For future reference, `malloc` returns a void pointer, so you should cast it to the pointer type you want.

Gjjjiang replied:
Done

dgobalak commented on onboarding/source/amb_light_service.c:
```diff
@@ -18,22 +26,91 @@ static void lightServiceTask(void * pvParameters);
+                                &event,
+                                0);
+        if (xReturned == pdPASS && event == MEASURE_LIGHT) {
+            adcStartConversion(adcREG1, adcGROUP1);
```
I'd recommend creating a helper function to do the ADC conversion. I.e. move this code to its own function.

Gjjjiang replied:
Done

dgobalak commented on onboarding/source/amb_light_service.c:
```diff
@@ -18,22 +26,91 @@ static void lightServiceTask(void * pvParameters);
     /* USER CODE BEGIN */
     // Send the event to the queue.
-    
+    xQueueSend(lightServiceQueueHandle,
```
Return 1 if `xQueueSend` is successful.

Gjjjiang replied:
Done

dgobalak commented on onboarding/source/amb_light_service.c:
```diff
@@ -18,22 +26,91 @@ static void lightServiceTask(void * pvParameters);
+            while(!adcIsConversionComplete(adcREG1, adcGROUP1));
+            adcGetSingleData(adcREG1, adcGROUP1, lightDataPtr);
+            char *msg = malloc(20 * sizeof(char));
+            sprintf(msg, "light level: %d", lightData.value);
```
It's a good idea to print a new line after each measurement so you can differentiate between them.

Gjjjiang replied:
Done

dgobalak commented on onboarding/source/amb_light_service.c:
```diff
@@ -6,7 +6,15 @@
+#include "os_task.h"
+#include "os_queue.h"
 
+void adcStartConversion_selChn(adcBASE_t *adc, unsigned channel, unsigned fifo_size, unsigned group);
```
It's clearer if you write the types as `unsigned int` instead of `unsigned`. It's even better if you use the appropriate fixed-size integer types for these arguments (such as `uint8_t` or `uint32_t`).

Gjjjiang replied:
Done

dgobalak commented on onboarding/source/controller.c:
```diff
@@ -70,15 +76,15 @@ static void controllerTask(void * pvParameters) {
         /* USER CODE BEGIN */
         // Deal with error when initializing light service task and/or queue
-
+        sciPrintText(scilinREG, "Error", 6);
```
Add a `\n` at the end of the error message.

Gjjjiang replied:
Done

dgobalak commented on onboarding/source/controller.c:
```diff
@@ -94,7 +100,9 @@ static void ledTimerCallback(TimerHandle_t xTimer) {
-
+    ASSERT(xTimer != NULL);
+    
+    sendToLightServiceQueue(MEASURE_LIGHT);
```
Print an error message if the `sendToLightServiceQueue` failed

Gjjjiang replied:
Done

dgobalak commented on onboarding/source/amb_light_service.c:
```diff
@@ -18,22 +26,91 @@ static void lightServiceTask(void * pvParameters);
+                                &event,
+                                0);
+        if (xReturned == pdPASS && event == MEASURE_LIGHT) {
+            adcStartConversion(adcREG1, adcGROUP1);
```
No need to call both `adcStartConversion` and `adcStartConversion_selChn`. It looks like you're following the example code by TI, so I'd just stick with `adcStartConversion_selChn`.

Gjjjiang replied:
Done

dgobalak commented on drivers/source/serial_io.c:
```diff
@@ -24,7 +24,9 @@ void sciMutexInit(void) {
     }
```
There should be some modified files in `hal` once you complete the HALCoGen steps in the challenge

Gjjjiang replied:
Oh, I just didn't commit these files.

dgobalak reviewed (changes requested):
Looks good so far. Just a few changes are needed.

Gjjjiang committed:
added hal/* files,
updated printed text,
added more error checking and messages,
added getADCConversion helper function,
removed dynamic memory allocation.

dgobalak commented on onboarding/source/amb_light_service.c:
```diff
@@ -6,7 +6,19 @@
+#include <stdio.h>
+
+void adcStartConversion_selChn(adcBASE_t *adc, uint8_t channel, uint8_t fifo_size, uint8_t group);
+void adcGetSingleData(adcBASE_t *adc, unsigned group, adcData_t *data);
```
Change this `unsigned` to the appropriate fixed-size integer type as well.

Gjjjiang replied:
Done

dgobalak commented on onboarding/source/amb_light_service.c:
```diff
@@ -6,7 +6,19 @@
+
+void adcStartConversion_selChn(adcBASE_t *adc, uint8_t channel, uint8_t fifo_size, uint8_t group);
+void adcGetSingleData(adcBASE_t *adc, unsigned group, adcData_t *data);
+uint16_t getADCConversion();
```
It'd make more sense to call this function something like `getLightValue`.

Gjjjiang replied:
Done

dgobalak commented on onboarding/source/amb_light_service.c:
```diff
@@ -6,7 +6,19 @@
+
+void adcStartConversion_selChn(adcBASE_t *adc, uint8_t channel, uint8_t fifo_size, uint8_t group);
+void adcGetSingleData(adcBASE_t *adc, unsigned group, adcData_t *data);
+uint16_t getADCConversion();
```
Since this function is only intended to be used by this source file, declare it a `static` function

Gjjjiang replied:
Done

dgobalak commented on onboarding/source/amb_light_service.c:
```diff
@@ -6,7 +6,19 @@
+
+void adcStartConversion_selChn(adcBASE_t *adc, uint8_t channel, uint8_t fifo_size, uint8_t group);
+void adcGetSingleData(adcBASE_t *adc, unsigned group, adcData_t *data);
+uint16_t getADCConversion();
```
In C, when a function takes no arguments, you should write `void` as the argument. For example, something like `uint16_t getADCConversion(void);`

Take a look at this: https://stackoverflow.com/questions/693788/is-it-better-to-use-c-void-arguments-void-foovoid-or-not-void-foo

Gjjjiang replied:
Done

dgobalak commented on onboarding/source/amb_light_service.c:
```diff
@@ -18,22 +30,101 @@ static void lightServiceTask(void * pvParameters);
+    */
+}
+
+uint16_t getADCConversion(){
```
Write `void` as the argument

Gjjjiang replied:
Done

Gjjjiang committed:
Changed getADCConvertion to static void getLightValue(void)
updated unsigned to uint8_t

dgobalak commented:
Great job! I'll close the PR. I'd say you're ready to take on a task. 

If you want, you can try flashing this program onto the RM46 LaunchPad at the weekly meeting or work session.
